MFEM Complex AuxKernels

Heinrich-BR opened the pull request:
Closes #31961

Heinrich-BR committed:
Added MFEMComplexAuxKernel and MFEMComplexGradAux

Heinrich-BR committed:
Added MFEMComplexSumAux

Heinrich-BR committed:
Added MFEMComplexCurlAux (closes #31962)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3399808/), step Clang format on c194e31 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/31962/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/31962/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 8738181fdd53615c94c02091e1ac1dda5d5cb226`

Heinrich-BR committed:
Added MFEMComplexDivAux

Heinrich-BR committed:
Added some simplifying methods

Heinrich-BR committed:
Added MFEMComplexCrossProductAux

Heinrich-BR committed:
Added MFEMComplexDotProductAux

Heinrich-BR committed:
Added MFEMComplexScalarProjectionAux

Added MFEMComplexVectorProjectionAux

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3399946/), step Clang format on 9f88b41 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/31962/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/31962/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 8738181fdd53615c94c02091e1ac1dda5d5cb226`

Heinrich-BR committed:
Add complex auxkernel docs

Linting

Heinrich-BR committed:
Add ComplexSumAux test

Update sign on complex dot product

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3400214/), step Results summary on 690e929 wanted to post the following:

## Framework test summary

Compared against 8738181 in job [civet.inl.gov/job/3397332](https://civet.inl.gov/job/3397332).

No change
## Modules test summary

Compared against 8738181 in job [civet.inl.gov/job/3397332](https://civet.inl.gov/job/3397332).

No change

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/3400147/), step Docs: sync website on 690e929 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/31962/site)

This comment will be updated on new commits.

Heinrich-BR committed:
Update docs and fix complexScale segfault

Fix complexScale

Heinrich-BR committed:
Add MFEMComplex[Grad,Curl,Div]Aux test

Add more complex auxkernel tests

Heinrich-BR committed:
Split up tests

Linting

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3410056/), step Clang format on 116318b wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/31962/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/31962/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format ecf7a09933f550b625784f9d5a71d20985022261`

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3414701/), step Results summary on 33246f4 wanted to post the following:

## Framework test summary

Compared against c1e94d1 in job [civet.inl.gov/job/3413879](https://civet.inl.gov/job/3413879).

No change
## Modules test summary

Compared against c1e94d1 in job [civet.inl.gov/job/3413879](https://civet.inl.gov/job/3413879).

No change

Heinrich-BR committed:
Renamed complex dot product

Heinrich-BR committed:
Renamed complex cross product

Linting

Heinrich-BR committed:
Rename doc file

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3422176/), step Results summary on f022f01 wanted to post the following:

## Framework test summary

Compared against 82bb60c in job [civet.inl.gov/job/3419953](https://civet.inl.gov/job/3419953).

No change
## Modules test summary

Compared against 82bb60c in job [civet.inl.gov/job/3419953](https://civet.inl.gov/job/3419953).

No change

nmnobre commented on framework/doc/content/source/mfem/auxkernels/MFEMComplexDivAux.md:
```diff
@@ -0,0 +1,33 @@
+
+where $\vec u \in H(\mathrm{div})$, $v \in L^2$ and $\lambda$ is a complex scalar constant.
+
+## Example Input File Syntax
```
```suggestion
## Input File Syntax
```
We chose this phrasing because we don't have actual examples, that right?

nmnobre commented on framework/src/mfem/auxkernels/MFEMHermitianInnerProductAux.C:
```diff
@@ -0,0 +1,88 @@
+//* This file is part of the MOOSE framework
```
Can we rebase and reimplement both the inner and external products in light of recent changes to their real counterparts?

nmnobre reviewed (changes requested):
Looking good, just a few tweaks.

Heinrich-BR committed:
Apply suggestions from code review

Co-authored-by: Nuno Nobre <nuno.nobre@stfc.ac.uk>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3426454/), step Clang format on 0fb43ea wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/31962/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/31962/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d67480e20140b67687fb556189c98a078fb0e3e1`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3426742/), step Versioner verify on ab378e0 wanted to post the following:

### Versioner templates

Found 14 templates, 0 failed


### Versioner influential files

Found 58 influential files, 1 changed, 0 added, 0 removed

| package   | status   | file                   |
|-----------|----------|------------------------|
| wasp      | CHANGE   | framework/contrib/wasp |

### Versioner versions

Found 9 packages, 2 changed, 2 failed

| package   | status    | hash    | to hash   | version                    | to version                 |
|-----------|-----------|---------|-----------|----------------------------|----------------------------|
| wasp      | NEED BUMP | 0320f16 | f278c09   | 2025.09.19_02960f1 build 0 | 2025.09.19_02960f1 build 0 |
| moose-dev | NEED BUMP | a20764f | fc18d70   | 2025.11.24                 | 2025.11.24                 |

Verification failed.

Heinrich-BR committed:
Renamings and small fixes

Heinrich-BR committed:
Removed unnecessary summary sections from complex auxkernel docs

nmnobre commented on framework/doc/content/source/mfem/auxkernels/MFEMComplexExteriorProductAux.md:
```diff
@@ -0,0 +1,25 @@
+# MFEMComplexExteriorProductAux
```
This file is based on an old version of `MFEMInnerProductAux.md`, the new version improves the language and tex notation, can we use that instead?

nmnobre commented on framework/include/mfem/auxkernels/MFEMComplexVectorProjectionAux.h:
```diff
@@ -0,0 +1,36 @@
+protected:
+  /// Reference to source coefficient.
+  mfem::VectorCoefficient & _vec_coef_real;
+  mfem::VectorCoefficient & _vec_coef_imag;
```
Likewise here.

nmnobre commented on framework/src/mfem/auxkernels/MFEMComplexCurlAux.C:
```diff
@@ -0,0 +1,58 @@
+    _source_var(*getMFEMProblem().getProblemData().cmplx_gridfunctions.Get(_source_var_name)),
+    _scale_factor_real(getParam<mfem::real_t>("scale_factor_real")),
+    _scale_factor_imag(getParam<mfem::real_t>("scale_factor_imag")),
+    _curl(_source_var.real().ParFESpace(), _result_var.real().ParFESpace())
```
Also applies to the div and grad auxkernels.

nmnobre commented on framework/include/mfem/auxkernels/MFEMComplexInnerProductAux.h:
```diff
@@ -0,0 +1,63 @@
+ *   - Currently supports only interior DOFs (no shared/constrained DOFs).
+ *   - Enforces 3D: mesh dimension and all involved vdim must be 3.
+ *   - Takes into account Hermitian conjugation when computing the dot product.
+ */
```
Edit as per `MFEMInnerProductAux.h` for correct math formatting and updated notes.

nmnobre commented on framework/include/mfem/auxkernels/MFEMComplexExteriorProductAux.h:
```diff
@@ -0,0 +1,63 @@
+//* This file is part of the MOOSE framework
```
See notes for `MFEMComplexInnerProductAux.h` and apply the same changes.

nmnobre commented on framework/src/mfem/auxkernels/MFEMComplexInnerProductAux.C:
```diff
@@ -0,0 +1,87 @@
+
+#include "MFEMComplexInnerProductAux.h"
+#include "MFEMProblem.h"
+#include "mfem.hpp"
```
```suggestion
```

Remove as per new `MFEMInnerProductAux.C`.

nmnobre commented on framework/src/mfem/auxkernels/MFEMComplexExteriorProductAux.C:
```diff
@@ -0,0 +1,91 @@
+  mfem::ParFiniteElementSpace * fes = _result_var.ParFESpace();
+  const int mesh_dim = fes->GetMesh()->Dimension();
+
+  // Enforce 3D cross product
```
Update as per `MFEMCrossProductAux.C`.

Heinrich-BR committed:
Apply suggestions from code review

Co-authored-by: Nuno Nobre <nuno.nobre@stfc.ac.uk>

Review suggestions

PR reviews

Fix MFEMAuxKernelTest

Style change

Linting

Linting

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3466821/), step Clang format on 4ec7bad wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/31962/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/31962/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format fb28c060e20cad83d698401c3d0f7b575de35980`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3466915/), step Clang format on 59637d7 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/31962/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/31962/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format fb28c060e20cad83d698401c3d0f7b575de35980`

nmnobre committed:
Final tweaks for complex auxkernel implementations

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3469468/), step Results summary on a620898 wanted to post the following:

## Framework test summary

Compared against fb28c06 in job [civet.inl.gov/job/3464837](https://civet.inl.gov/job/3464837).

No change
## Modules test summary

Compared against fb28c06 in job [civet.inl.gov/job/3464837](https://civet.inl.gov/job/3464837).

No change

nmnobre reviewed (approved):
Thanks Henrique, I think this is looking extremely good now! :)
